fixed issue to where screen reader reads all buttons when opening datepicker

DBowen33 opened the pull request:
removed container=true property from Semantics wrapper of datepicker dialog. This prevents the screen reader from reading every button in the datepicker dialog whenever a user opens it, reducing SR noise.

Before: https://screencast.googleplex.com/cast/Njc0Mzc1MDA5MTk5NzE4NHw2NTU5ODI4YS0xNA
After: https://screencast.googleplex.com/cast/NTYxMDUwOTIxMzYzMDQ2NHxjOWQ2M2YzNy1hYQ

fixes b/345297872

NOTE: Please let me know if a test is needed for this, since it is only deleting code.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

DBowen33 committed:
fixed issue to where screen reader reads all button when opening datepicker dialog

DBowen33 committed:
revert unneccessary fix

hangyujin commented on packages/flutter/lib/src/material/date_picker.dart:
```diff
@@ -927,7 +927,6 @@ class _DatePickerHeader extends StatelessWidget {
       case Orientation.portrait:
         return Semantics(
-          container: true,
           child: SizedBox(
```
if `container: true,` is removed,  `Semantics` wrapper here can also be removed since it added nothing

DBowen33 replied:
removed Semantics wrapper

hangyujin commented on packages/flutter/lib/src/material/date_picker.dart:
```diff
@@ -962,7 +961,6 @@ class _DatePickerHeader extends StatelessWidget {
         );
       case Orientation.landscape:
         return Semantics(
```
ditto

DBowen33 replied:
removed Semantics wrapper

hangyujin commented:
Yes a PR like this need a test, because the behavior is changed.

DBowen33 committed:
remove semantics wrapper

DBowen33 committed:
remove test thats not working

DBowen33 committed:
format

DBowen33 committed:
add test

DBowen33 commented:
Added test

DBowen33 committed:
Trigger Build

DBowen33 committed:
trigger build

hangyujin commented:
Removing it will also cause mobile a11y behavior to change.

Semantics `container` here controls if the `date picker header` is a node in the semantics tree, This line was added in https://github.com/flutter/flutter/pull/143117  to solve issue https://github.com/flutter/flutter/issues/141992. The date picker header being a node is consistent with native a11y  behavior.  So I think we should keep the Semantics `container` for the date picker header.

DBowen33 committed:
redo fix

DBowen33 committed:
remove old code

DBowen33 committed:
fix tests for adding semantics back to calendar year mode toggle

DBowen33 committed:
fix test

DBowen33 commented:
@hangyujin thank you for the feedback. I went ahead and reverted the change and instead added a container to the  month toggle section. This solves the behavior of reading all of the interactive elements in the datepicker dialog and instead just reads the heading when user opens datepicker dialog and focus goes on it.

Also added `explicitChildNodes: true` so that the previous and next buttons don't get grouped into the Semantics node of the month selector toggle button on the top. This prevents the SR from reading out the next and previous buttons when the user is focused on selecting a new day.

Also removed `excludeSemantics: true` from the datepicker toggle button because I noticed that when user is focused on that button, the Semantics information was not being read out to the user. Not quite sure why that was added on that.

DBowen33 commented:
I'm also having some trouble creating a test case to reflect these new changes, if you have any ideas on what direction to take for this it would be greatly appreciated! :)

hangyujin commented:
The new changes looks good to me! 

For tests:
You tests already reflect the `excludeSemantics` changes.

As for `container: true` , You can either use `final SemanticsTester semantics = SemanticsTester(tester);` and test the semantics tree changes or use `tester.getSemantics()` to find a semantics node and test it.

From the semantics tree it looks like  adding `container: true` will remove the `isFocusable` flag in the date picker SemanticsNode.  You can add a test to reflect that.

DBowen33 committed:
add tests

DBowen33 committed:
change test name

DBowen33 commented:
> The new changes looks good to me!
> 
> For tests: You tests already reflect the `excludeSemantics` changes.
> 
> As for `container: true` , You can either use `final SemanticsTester semantics = SemanticsTester(tester);` and test the semantics tree changes or use `tester.getSemantics()` to find a semantics node and test it.
> 
> I tried on my mobile phone and from the semantics tree it looks like adding `container: true` in the date picker Semantics will remove the `isFocusable` flag in the date picker SemanticsNode. You can add a test to reflect that.

Thanks Hangyu! So I went ahead and added a test to test the semantics tree of the datepicker. I'm not sure if there is a way to have a more succinct test where I don't have to test all of the SemanticsNode children involved in the datepicker, but rather just grab the SemanticsNode I want (datepicker SemanticsNode). If there is a way to use `.find` to find that particular node datepicker node where the `isFocusable`  flag is not added anymore, please let me know!

DBowen33 committed:
make test cleaner

DBowen33 committed:
remove import

DBowen33 commented:
> The new changes looks good to me!
> 
> For tests: You tests already reflect the `excludeSemantics` changes.
> 
> As for `container: true` , You can either use `final SemanticsTester semantics = SemanticsTester(tester);` and test the semantics tree changes or use `tester.getSemantics()` to find a semantics node and test it.
> 
> I tried on my mobile phone and from the semantics tree it looks like adding `container: true` in the date picker Semantics will remove the `isFocusable` flag in the date picker SemanticsNode. You can add a test to reflect that.

Added test, thank you for the tips!

hannah-hyj reviewed (approved):
LGTM, thanks!
